language/move-binary-format/src/deserializer.rs

```diff
@@ -12,29 +12,46 @@ use std::{collections::HashSet, convert::TryInto, io::Read};
 impl CompiledScript {
     /// Deserializes a &[u8] slice into a `CompiledScript` instance.
     pub fn deserialize(binary: &[u8]) -> BinaryLoaderResult<Self> {
-        let script = CompiledScript::deserialize_no_check_bounds(binary)?;
+        Self::deserialize_with_max_version(binary, VERSION_MAX)
+    }
+
+    /// Deserializes a &[u8] slice into a `CompiledScript` instance.
+    pub fn deserialize_with_max_version(
+        binary: &[u8],
+        max_binary_format_version: u32,
+    ) -> BinaryLoaderResult<Self> {
+        let script = deserialize_compiled_script(binary, max_binary_format_version)?;
```

Comment:
Maybe we can provide a function to auto-convert the new binary format to the old format.

If a new binary module does not contain the new Bytecodes, it will be converted to the old format, like  https://github.com/starcoinorg/starcoin/blob/master/vm/compiler/src/bytecode_transpose.rs

This can let the upgrade progress more gracefully.

Reply:
This is an interesting idea I think we should definitely explore! Conceptually I see this as adding an additional compatibility mode or changing `max_binary_format_version` into an enum:
```
enum BinaryFormatCompatibilityOption {
    Upto(u32),
    CompatibleWIth(u32),
}
```

However I'd suggest we try not to squeeze this into this PR since 
- This may require some additional discussions and design work 
- The relaxed requirement does NOT actually address the VM's immediate backward compatibility concern. Let me explain:
  - Right now, existing validators run the old software, which outright rejects binary format v6
  - You want your new software to have the ability to
    - Maintain full backward compatibility with the old one, meaning that they need to have a mode that would allow them to fully reject format v6. Otherwise validators running old and new software can go out of sync during an upgrade.
    - Accept format v6 with some configuration changes
  - This means, the current behavior I'm implementing in this PR (the Upto variant in the enum above) needs to be an option even with the addition of the more relaxed mode